ROX-16696 - Delegated Registry Config Storage / APIs

dcaravel opened the pull request:
## Description

This is a part of larger epic [ROX-15447](https://issues.redhat.com/browse/ROX-15447) to enable 'on-prem registry scanning'. It introduces a new `Delegated Registry Config` that will eventually be used to route scan requests to an appropriate destination.

To persist the config a new PostgreSQL table was created, along with a REST API for retrieving/updating the config.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [x] Determined and documented upgrade steps
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Unit tested along with deployed to GKE cluster and tested with `curl`. Verified config persisted by updating config, restarting central, and observing config changes were retrieved

```sh
export BASE=https://<central>/v1/delegatedregistryconfig

## get current config
curl -k -u "$U:$P" "$BASE"

{"enabledFor":"NONE","defaultClusterId":"","registries":[]}

## get eligible clusters
curl -k -u "$U:$P" "$BASE/clusters"

{"clusters":[{"id":"71bbbfd7-ae66-4bf4-86c1-17ff9fa983f1","name":"local-scan","isValid":true}]}

## update config
curl -k -u "$U:$P" "$BASE" -X PUT \
  --data-raw '{"enabledFor":"ALL", "defaultClusterId": "71bbbfd7-ae66-4bf4-86c1-17ff9fa983f1"}'

{"enabledFor":"ALL","defaultClusterId":"71bbbfd7-ae66-4bf4-86c1-17ff9fa983f1","registries":[]}

## get current config
curl -k -u "$U:$P" "$BASE"

{"enabledFor":"ALL","defaultClusterId":"71bbbfd7-ae66-4bf4-86c1-17ff9fa983f1","registries":[]}

## restart central and get
kubectl delete pods -n stackrox -lapp=central
curl -k -u "$U:$P" "$BASE"

{"enabledFor":"ALL","defaultClusterId":"71bbbfd7-ae66-4bf4-86c1-17ff9fa983f1","registries":[]}

## test invalid enum
curl -k -i -u "$U:$P" "$BASE" -X PUT --data-raw '{"enabledFor":"INVALID"}'

{"error":"unknown value \"\\\"INVALID\\\"\" for enum storage.DelegatedRegistryConfig.EnabledFor","code":3,"message":"unknown value \"\\\"INVALID\\\"\" for enum storage.DelegatedRegistryConfig.EnabledFor","details":[]}

## invalid cluster
curl -k -i -u "$U:$P" "$BASE" -X PUT \
  --data-raw '{"enabledFor":"ALL", "defaultClusterId": "hello"}'

{"error":"Validation error: default cluster \"hello\" is not a valid cluster: invalid arguments","code":3,"message":"Validation error: default cluster \"hello\" is not a valid cluster: invalid arguments","details":[]}
```

dcaravel committed:
initial delegated registry config storage/api

dcaravel committed:
add drc api

dcaravel committed:
wip clusters, api validation

dcaravel committed:
tests, validation

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel committed:
update dele cluster flag comment

dcaravel committed:
remove debug log, add datastore missing msg

dcaravel committed:
cleanup unused stuff

dcaravel committed:
adjust name to avoid secrets swagger gen change

dcaravel committed:
add tests, fix style errors

dashrews78 commented on proto/api/v1/delegated_registry_config_service.proto:
```diff
@@ -0,0 +1,50 @@
+
+import weak "google/api/annotations.proto";
+import "api/v1/empty.proto";
+import "storage/delegated_registry_config.proto";
```
I highly recommend not including the storage proto in the api proto.  That introduces a handcuff to updating the storage later.

dcaravel replied:
I have heard that in conversation - wasn't sure if that was a 'big bang' plan or just going forward, I'll swap it out for something specific for the API

connorgorman replied:
👍

dcaravel replied:
Believe this is addressed in the latest commit(s), let me know if thats no what you had in mind.

dashrews78 commented on central/delegatedregistryconfig/datastore/singleton.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func initialize() {
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
```
Probably not a question for you per se, but at what point can we stop putting these  in the code?

dcaravel replied:
I'm OK doing that now, was following what I saw elsewhere for consistency.  I see that the variable is permanently `true` so might as well remove it. For this too wasn't sure if there was a planned 'purge' of all the conditions throughout the codebase.

dcaravel replied:
Removed in latest commit

connorgorman commented on central/delegatedregistryconfig/datastore/datastore.go:
```diff
@@ -0,0 +1,56 @@
+
+// GetConfig returns Central's delegated registry config
+func (d *datastoreImpl) GetConfig(ctx context.Context) (*storage.DelegatedRegistryConfig, error) {
+	if ok, err := administrationSAC.ReadAllowed(ctx); err != nil {
```
@rhybrillou Is this necessary here anymore?

connorgorman commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,227 @@
+		return nil
+	}
+
+	errorList := errorhelpers.NewErrorList("Validation")
```
There is a new type of error joining which is in go 1.20: https://go.dev/blog/go1.20 where you can join errors so that errors.Is and errors.As still works

dcaravel replied:
Nice! I took a look - in this particular workflow I'm currently not inspecting the type of errors - is there a scenario I'm missing that will benefit from `errors.Join`?

connorgorman replied:
namely the one in which I want to deprecate NewErrorList :)

dcaravel replied:
ahh, didn't realize `errorList` was on the way out - ok I'll remove `errorList` in favor of a `[]error` slice and `errors.Join`

dcaravel replied:
take a look at the latest commit, that what you had in mind?

connorgorman commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,227 @@
+	errorList := errorhelpers.NewErrorList("Validation")
+	// validate the default cluster id
+	if config.DefaultClusterId == "" {
+		errorList.AddStrings("defaultClusterId required if enabledFor is not NONE")
```
typically I would just write this out in plan english and not using the camel case style

dcaravel replied:
fixed in latest commit

connorgorman commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,227 @@
+
+		// if a cluster id was provided, check if its valid
+		if r.ClusterId != "" && !validClusters.Contains(r.ClusterId) {
+			errorList.AddStrings(fmt.Sprintf("cluster %q is not valid at index %v", r.ClusterId, i))
```
i would omit the at index portion

dcaravel replied:
My thought was to make it easier on debugging due to the possibility that multiple 'registries' could be pointing to the same invalid cluster, and the UI would be showing the cluster name not ID. Thinking about it more there are other ways the invalid line items could be identified, so I'll remove the index.

dcaravel replied:
Removed in latest commit

connorgorman commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,227 @@
+		return nil, nil
+	}
+
+	var res []*v1.DelegatedRegistryCluster
```
since we know that the size of res will equal the size of clusters, you can presize this slice which makes appending faster because it won't resize the slice

res := make([]*v1.DelegatedRegistryCluster, 0, len(clusters))

dcaravel replied:
good catch, missed this one - originally started with not including 'invalid' clusters, so the size wasn't known ahead of time - will make the change

dcaravel replied:
fixed, initialized length instead of capacity, from what I understand thats slightly faster

connorgorman reviewed (changes requested):
see comments

dcaravel commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,227 @@
+
+// GetConfig returns Central's delegated registry config
+func (s *serviceImpl) GetConfig(ctx context.Context, _ *v1.Empty) (*storage.DelegatedRegistryConfig, error) {
+	if s.dataStore == nil {
```
It was related to postgres not being enabled - since the postgres enabled flag will now always be true believe its unnecessary - i'll remove it

dcaravel replied:
Removed in latest commit

dcaravel committed:
cleanup tests, make code review changes

- separate api and storage protos
- remove references to errorList
- remove datastore checks for postgres enabled
- create util to convert from storage to api protos

dcaravel committed:
fix test case msg check

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel committed:
add ticket num to TODO

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel committed:
cleanup converster tests and naming

connorgorman commented on central/delegatedregistryconfig/datastore/datastore.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// GetConfig returns Central's delegated registry config
+func (d *datastoreImpl) GetConfig(ctx context.Context) (*storage.DelegatedRegistryConfig, error) {
```
In general I'm a fan of passing back the bool and not using nil as the flag that the value didn't exist

dashrews78 reviewed (approved):
my concerns were addressed.  LGTM.

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel committed:
use bool instead of nil for exists test

RTann commented on proto/api/v1/delegated_registry_config_service.proto:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // PutConfig updates the stored delegated registry configuration
+    rpc PutConfig(DelegatedRegistryConfig) returns (DelegatedRegistryConfig) {
```
nit: I think `UpdateConfig` may make more sense from an RPC standpoint (and separates it from the HTTP method)

dcaravel replied:
I agree, saw `Put` being used for the related services (cluster, config, etc.) and felt keep things similar

RTann commented on proto/api/v1/delegated_registry_config_service.proto:
```diff
@@ -0,0 +1,80 @@
+    // if enabledFor is NONE
+    string default_cluster_id    = 2;
+
+    // list of registries that has different outcome depending on value of enabled_for, allows
```
`list of registries that has different outcome` not sure I follow this

dcaravel replied:
I'll see if I can wordsmith that better - the registries in the list is interpreted differently depending on use case (ie: ad-hoc vs. informer event) and/or the value of `enabled for`

RTann commented on proto/api/v1/delegated_registry_config_service.proto:
```diff
@@ -0,0 +1,80 @@
+    // determines if delegation is enabled for no registries, all registries, or specific registries
+    EnabledFor enabled_for       = 1;
+
+    // the default cluster to delegate ad-hoc requests to if no match found in []registries, not used
```
nit: `[]` isn't needed

RTann commented on central/delegatedregistryconfig/convert/convert.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	// defaults to 0 (NONE) if not found in map
+	enabledFor := v1.DelegatedRegistryConfig_EnabledFor_value[from.EnabledFor.String()]
```
I always opt for manual `switch`, but this works as long as we are careful. This can be dangerous if we were to change the name of the enum (I believe that's allowed and compatible in proto as long as the tag is not changed)

RTann replied:
I'm thinking let's make a function to do this conversion ourselves for safety

dcaravel replied:
Wouldn't changing the name of the enum cause a compile time error? The map would have a different name in that case unless I'm missing something.  What did you have in mind for a function? Each of the types could potentially have a different set of names to values (`int32`'s) in the future, the map seemed to be the safest option in that case.  Unit tests would fail as well if the values changed to something unexpected.

RTann replied:
I was wondering what may happen if `from.EnabledFor.String()` is no longer something which matches `NONE`, `ALL`, or `SPECIFIC`. For example, another option is added for this proto but not the other. What will this default to? Or, what if we change the API's name but not the storage proto's name? We may default to something unexpected, which we would catch with a compile-time error instead. Let me know if that makes sense

dcaravel replied:
> I was wondering what may happen if from.EnabledFor.String() is no longer something which matches NONE, ALL, or SPECIFIC. For example, another option is added for this proto but not the other. What will this default to?

I believe that use case is captured in the [convert_test.go](https://github.com/stackrox/stackrox/blob/dc/ROX-16696-delegated-reg-config-storage-api/central/delegatedregistryconfig/convert/convert_test.go#L56) - For good measure however, the below test shows that an entry not found in the map will default to `0`

```go
func TestTemp(t *testing.T) {
	enabledFor := v1.DelegatedRegistryConfig_EnabledFor_value["HAPPY TUESDAY!"]
	fmt.Println(enabledFor)
}
```
Output:
```
=== RUN   TestTemp
0
--- PASS: TestTemp (0.00s)
```

> Or, what if we change the API's name but not the storage proto's name? We may default to something unexpected, which we would catch with a compile-time error instead. Let me know if that makes sense

If we change one name and not the other the tests in `convert_test.go` will fail.

For good measure changed `proto/api/v1/delegated_registry_config_service.proto` to have:

```proto
...
    enum EnabledFor {
        HELLO     = 0;
        WORLD      = 1;
        YIPPIE = 2;
    }
...
```
And the test case no longer compiles:
![image](https://github.com/stackrox/stackrox/assets/119438707/ebeb8626-2798-44c6-8956-a00c81ab309b)

RTann replied:
as long as it is captured somewhere by compile time errors, I'm ok with this

RTann commented on central/delegatedregistryconfig/convert/convert.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	// defaults to 0 (NONE) if not found in map
+	enabledFor := storage.DelegatedRegistryConfig_EnabledFor_value[from.EnabledFor.String()]
```
same as the previous comment

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+func (s *serviceImpl) GetConfig(ctx context.Context, _ *v1.Empty) (*v1.DelegatedRegistryConfig, error) {
+	config, exists, err := s.dataStore.GetConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving config %w", err)
```
nit: [`status.Errorf`](https://pkg.go.dev/google.golang.org/grpc/status#Errorf)

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+func (s *serviceImpl) GetClusters(ctx context.Context, _ *v1.Empty) (*v1.DelegatedRegistryClustersResponse, error) {
+	clusters, err := s.getClusters(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving clusters %w", err)
```
I was going to to mention `status.Error` here, but I see it is used below. Curious, why not here, too?

dcaravel replied:
returning an error defaults to a `500` which is whats desired in those cases, was only using it when needed to change the status code to a `400` , was modeling this off of the config & cluster services which split usage as well

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+
+func (s *serviceImpl) validate(ctx context.Context, config *v1.DelegatedRegistryConfig) error {
+	if config == nil {
+		// this block not reachable via GRPC-gateway invocations
```
is it part of the path?

dcaravel replied:
the only current known consumer of this endpoint will be the UI via the grpc-gateway, so not in that path in that regards, I believe it's possible a GRPC consumer could reach that path.  I can remove that comment if its misleading

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+		return fmt.Errorf("unable to validate config %w", err)
+	}
+
+	var errorList []error
```
`errorhelpers.NewErrorList` (our version of `errors.Join` before go1.20. A little bit of a cleaner usage IMO)

dcaravel replied:
Thats what I had used originally, changed it to `errors.Join` in response to the code review comments above. While I agree the helper is sugar, the direction of relying on the standard library where it makes sense seems like the right decision long term.

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+
+// getClusters returns all clusters annotated with a flag indicating if cluster is valid
+// for use as a delegation target. All clusters are returned instead of just valid clusters
+// so that a consumer (ie: the UI) can show the friendly name of clusters that may no longer
```
I think this part may be a little confusing, and I think you can just remove the explanation to resolve that.

dcaravel replied:
I think its important to explain the why the valid flag is there if its questioned in the future. Perhaps I can re-word it to make it more clear

RTann commented on central/delegatedregistryconfig/service/service.go:
```diff
@@ -0,0 +1,213 @@
+
+// GetClusters returns the list of clusters (id + name) that are eligible for delegating scanning
+// requests (ie: only clusters with scanners that understand the delegated registry config)
+func (s *serviceImpl) GetClusters(ctx context.Context, _ *v1.Empty) (*v1.DelegatedRegistryClustersResponse, error) {
```
Why use this vs the ClusterService?

dcaravel replied:
I am using the datastore behind ClusterService, so technically using the same back-end. The reason for using a separate method is to make it so consumers (ie: the UI) do not have to bake business logic into the presentation tier. If the conditions which determine a cluster is valid for use as a 'delegation target' ever change, the UI and consumers will not need to.  In the UX/UI discussions, this was the preferred approach, same thing is done as part of network graph.

RTann reviewed (commented):
a few comments to consider before merging

dcaravel committed:
code review changes

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel commented on proto/api/v1/delegated_registry_config_service.proto:
```diff
@@ -14,39 +14,46 @@ message DelegatedRegistryClustersResponse {
 message DelegatedRegistryConfig {
     enum EnabledFor {
-        // delegate no scans
+        // Scan all images via central services except for images from the OCP integrated registry
```
In this particular PR no.  In the subsequent PR the env var is removed: https://github.com/stackrox/stackrox/pull/6005, so no the actions of the `DelegatedRegistryConfig` are not related to the MVP env var

dcaravel committed:
Merge remote-tracking branch 'origin/master' into dc/ROX-16696-delegated-reg-config-storage-api

dcaravel committed:
update godoc to match method name
